documentation/docs/06-form-actions.md

```diff
@@ -0,0 +1,254 @@
+
+```js
+/// file: src/routes/login/+page.server.js
+
+// @filename: ambient.d.ts
+declare global {
+	const db: {
+		findUser: (name: string) => Promise<{
+			id: string;
+			username: string;
+			password: string;
+		}>
+	}
+}
+
+export {};
+
+// @filename: index.js
+// ---cut---
+import { invalid } from '@sveltejs/kit';
+
+/** @type {import('./$types').Actions} */
+export const actions = {
+	default: async ({ fields, setHeaders, url }) => {
+		const username = fields.get('username');
+		const password = fields.get('password');
+
+		const user = await db.findUser(username);
+
+		if (!user) {
+			throw invalid(403, { username }, {
+				username: 'No user with this username'
+			});
+		}
+
+		// ...
+	}
+};
+```
+
+```svelte
+/// file: src/routes/login/+page.svelte
+<script>
+	import { submitted } from '$app/stores';
+</script>
+
+<form action="?/addTodo" method="post">
+	<input type="text" name="username" value={$submitted?.values?.username} />
+	{#if $submitted?.errors?.username}
+		<span>{$submitted?.errors?.username}</span>
+	{/if}
+	<input type="password" name="password" />
+	<button>Login</button>
+</form>
+```
+
+## Success
+
+If everything is valid, an action can return a JSON object with data that is part of the JSON response in the case of a JavaScript fetch - it's discarded in case of a full page reload. Alternatively it can `throw` a `redirect` to redirect the user to another page.
+
+```js
+/// file: src/routes/login/+page.server.js
+import { redirect } from '@sveltejs/kit';
+
+/** @type {import('./$types').Actions} */
+export const actions = {
+	default: async ({ url }) => {
+		// ...
+
+		if (url.searchParams.get('redirectTo')) {
+			throw redirect(303, url.searchParams.get('redirectTo'));
+		} else {
+			return {
+				success: true
+			};
+		}
+	}
+};
+```
+
+## Progressive enhancement
+
+So far, all the code examples run native form submissions - that is, when the user pressed the submit button, the page is reloaded. It's good that this use case is supported since JavaScript may not be loaded all the time. When it is though, it might be a better user experience to use the powers JavaScript gives us to provide a better user experience - this is called progressive enhancement.
+
+First we need to ensure that the page is _not_ reloaded on submission. For this, we prevent the default behavior. Afterwards, we run our JavaScript code instead which does the form submission through `fetch` instead.
+
+```svelte
+/// file: src/routes/login/+page.svelte
+<script>
+	import { submitted } from '$app/stores';
+	import { invalidateAll, goto } from '$app/navigation';
+
+	async function login(event) {
+		event.preventDefault(); // prevent native form submission
```

Comment:
Why are we calling `preventDefault` again here?
The form already has `on:submit|preventDefault={login}`

Reply:
good catch, this should be deduplicated - not sure which way though. `|preventDefault` is more Svelte-like, but as a reader you may not have seen that before, and with `event.preventDefault` we can put an explanatory comment next to it - so I probably lean towards the latter.

Reply:
I feel `|preventDefault` looks clean and makes sense for docs to use more of svelte way.
But yeah we loose comment option.